notion-server/src/ethereum/mod.rs

```diff
@@ -226,36 +229,62 @@ impl Simulation {
-        if let Some(shard_chain) = self.shard_chains.get_mut(args.shard_chain_index as usize) {
-            let shard_block = ShardBlock::try_from(args.shard_block)?;
-
-            // TODO: Run each transaction (which will update the EE state for that shard)
-            // Questions to answer:
-            //   * What if the decoding of the base64 data string fails? Remove this transaction from the block?  Send back error value as result?
-            //   * What if executing the EE code fails with the given data? (Same options as above?)
-            // Example code from previous brainstorm:
-            //        let transactions = shard_block.transactions
-            //
-            //        for transaction in shard_block.transactions {
-            //            // This executes everything and presumably also updates the EE State on the shard
-            //            let ee = transaction.execution_environment;
-            //            let input_data = transaction.data;
-            //
-            //            let code = self.beacon_chain.get(ee);
-            //            let runtime = RootRuntime::new(&code, shard_ee_state_or_something_similar);
-            //            runtime.execute(input_data);
-            //        }
-
-            shard_chain.shard_blocks.push(shard_block);
-            Ok((shard_chain.shard_blocks.len() - 1) as u32)
-        } else {
-            Err(Error::OutOfBounds {
+    fn create_shard_block(&mut self, args: args::CreateShardBlock) -> Result<u32> {
+        let mut shard_chain = self
+            .shard_chains
+            .get_mut(args.shard_chain_index as usize)
+            .context(OutOfBounds {
                 message: format!("No shard chain exists at index: {}", args.shard_chain_index),
-            })
+            })?;
+
+        let mut transactions = Vec::with_capacity(args.shard_block.transactions.len());
+
+        for transaction in args.shard_block.transactions {
+            let execution_environment = self
+                .beacon_chain
+                .execution_environments
+                .get(transaction.ee_index as usize)
+                .context(OutOfBounds {
+                    message: format!(
+                        "No execution environment exists at index: {}",
+                        transaction.ee_index
+                    ),
+                })?;
+            let code = &execution_environment.wasm_code;
+
+            let pre_state = shard_chain
+                .execution_environment_state
+                .get(&EeIndex(transaction.ee_index))
+                .unwrap_or(&execution_environment.initial_shard_state);
+            let data = base64::decode(&transaction.base64_encoded_data).context(Decode)?;
+            let mut runtime = RootRuntime::new(&code, &data, pre_state.data);
+            let post_root = runtime.execute();
+            drop(runtime);
+
+            shard_chain.execution_environment_state.insert(
```

Comment:
Can this fail eg. if we are trying to insert into slot 3 and currently only have the EE state filled through slot 1?

Eg. Add 5 EEs to beacon chain, then try to run the 5th EE on a new shard chain -- I believe in current implementation that shard chain will have no state for any EEs, so not sure what the behavior is for indices 0-3 when we try to insert to index 4 in this vector.

Possible solutions: Make EE state (EEs on the beacon chain) an array with max size equal to MAX_NUMBER_OF_SHARDS, initialized to empty state for all shards, OR make sure a blank EE state is added at the appropriate index whenever a new EE is added and on initialization of a new Shard chain.

Reply:
I don't think this can fail. `shard_chain.execution_environment_state` is a `HashMap` not a `Vec`, which should avoid that problem.

Reply:
ah ok cool.